// v-calendar/lib/components/CustomTransition/index.ts
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('vue')) :
  typeof define === 'function' && define.amd ? define(['exports', 'vue'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.CustomTransition = {}, global.Vue));
}(this, (function (exports, vue) { 'use strict';

  var script = {
    name: 'CustomTransition',
    emits: ['before-enter', 'before-transition', 'after-enter', 'after-transition'],
    props: {
      name: String,
      appear: Boolean
    },
    computed: {
      name_: function name_() {
        return this.name || 'none';
      }
    },
    methods: {
      beforeEnter: function beforeEnter(el) {
        this.$emit('before-enter', el);
        this.$emit('before-transition', el);
      },
      afterEnter: function afterEnter(el) {
        this.$emit('after-enter', el);
        this.$emit('after-transition', el);
      }
    }
  };

  function render(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createBlock(vue.Transition, {
      name: $options.name_,
      appear: $props.appear,
      onBeforeEnter: $options.beforeEnter,
      onAfterEnter: $options.afterEnter
    }, {
      default: vue.withCtx(function () {
        return [vue.renderSlot(_ctx.$slots, "default")];
      }),
      _: 3
    }, 8, ["name", "appear", "onBeforeEnter", "onAfterEnter"]);
  }

  function styleInject(css, ref) {
    if ( ref === void 0 ) ref = {};
    var insertAt = ref.insertAt;

    if (!css || typeof document === 'undefined') { return; }

    var head = document.head || document.getElementsByTagName('head')[0];
    var style = document.createElement('style');
    style.type = 'text/css';

    if (insertAt === 'top') {
      if (head.firstChild) {
        head.insertBefore(style, head.firstChild);
      } else {
        head.appendChild(style);
      }
    } else {
      head.appendChild(style);
    }

    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }
  }

  var css_248z = ".none-enter-active,\r\n.none-leave-active {\r\n  transition-duration: 0s;\r\n}\r\n\r\n.fade-enter-active,\r\n.fade-leave-active,\r\n.slide-left-enter-active,\r\n.slide-left-leave-active,\r\n.slide-right-enter-active,\r\n.slide-right-leave-active,\r\n.slide-up-enter-active,\r\n.slide-up-leave-active,\r\n.slide-down-enter-active,\r\n.slide-down-leave-active,\r\n.slide-fade-enter-active,\r\n.slide-fade-leave-active {\r\n  transition: opacity var(--slide-duration) var(--slide-timing),\r\n    -webkit-transform var(--slide-duration) var(--slide-timing);\r\n  transition: transform var(--slide-duration) var(--slide-timing),\r\n    opacity var(--slide-duration) var(--slide-timing);\r\n  transition: transform var(--slide-duration) var(--slide-timing),\r\n    opacity var(--slide-duration) var(--slide-timing),\r\n    -webkit-transform var(--slide-duration) var(--slide-timing);\r\n  -webkit-backface-visibility: hidden;\r\n          backface-visibility: hidden;\r\n  pointer-events: none;\r\n}\r\n\r\n.none-leave-active,\r\n.fade-leave-active,\r\n.slide-left-leave-active,\r\n.slide-right-leave-active,\r\n.slide-up-leave-active,\r\n.slide-down-leave-active {\r\n  position: absolute !important;\r\n  width: 100%;\r\n}\r\n\r\n.none-enter-from,\r\n.none-leave-to,\r\n.fade-enter-from,\r\n.fade-leave-to,\r\n.slide-left-enter-from,\r\n.slide-left-leave-to,\r\n.slide-right-enter-from,\r\n.slide-right-leave-to,\r\n.slide-up-enter-from,\r\n.slide-up-leave-to,\r\n.slide-down-enter-from,\r\n.slide-down-leave-to,\r\n.slide-fade-enter-from,\r\n.slide-fade-leave-to {\r\n  opacity: 0;\r\n}\r\n\r\n.slide-left-enter-from,\r\n.slide-right-leave-to,\r\n.slide-fade-enter-from.direction-left,\r\n.slide-fade-leave-to.direction-left {\r\n  -webkit-transform: translateX(var(--slide-translate));\r\n          transform: translateX(var(--slide-translate));\r\n}\r\n\r\n.slide-right-enter-from,\r\n.slide-left-leave-to,\r\n.slide-fade-enter-from.direction-right,\r\n.slide-fade-leave-to.direction-right {\r\n  -webkit-transform: translateX(calc(-1 * var(--slide-translate)));\r\n          transform: translateX(calc(-1 * var(--slide-translate)));\r\n}\r\n\r\n.slide-up-enter-from,\r\n.slide-down-leave-to,\r\n.slide-fade-enter-from.direction-top,\r\n.slide-fade-leave-to.direction-top {\r\n  -webkit-transform: translateY(var(--slide-translate));\r\n          transform: translateY(var(--slide-translate));\r\n}\r\n\r\n.slide-down-enter-from,\r\n.slide-up-leave-to,\r\n.slide-fade-enter-from.direction-bottom,\r\n.slide-fade-leave-to.direction-bottom {\r\n  -webkit-transform: translateY(calc(-1 * var(--slide-translate)));\r\n          transform: translateY(calc(-1 * var(--slide-translate)));\r\n}\n";
  styleInject(css_248z);

  script.render = render;

  exports.default = script;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
