// v-calendar/lib/cjs/CalendarPane.vue_vue&type=style&index=0&lang-20f5d9e6.js
'use strict';var helpers=require('./helpers-fe8dbfcf.js'),vue=require('vue'),index=require('./index-f3a4fb5b.js'),CalendarNav=require('./CalendarNav-bee8fe94.js'),Popover_vue_vue_type_style_index_0_lang=require('./Popover.vue_vue&type=style&index=0&lang-b19ded44.js'),styleInject_es=require('./style-inject.es-06def3b0.js'),Grid_vue_vue_type_style_index_0_lang=require('./Grid.vue_vue&type=style&index=0&lang-c21310ab.js'),CalendarDay_vue_vue_type_style_index_0_lang=require('./CalendarDay.vue_vue&type=style&index=0&lang-9a74de40.js');var script = {
  name: 'CalendarPane',
  emits: ['update:page'],
  mixins: [index.j, index.f],
  inheritAttrs: false,
  render: function render() {
    var _this = this;

    // Header
    var header = this.safeSlot('header', this.page) || vue.h('div', {
      class: 'vc-header'
    }, [// Header title
    vue.h('div', {
      class: "vc-title-layout align-".concat(this.titlePosition)
    }, [vue.h('div', {
      class: 'vc-title-wrapper'
    }, [// Title content
    vue.h('div', helpers.m({
      class: 'vc-title'
    }, this.navPopoverEvents), [this.safeSlot('header-title', this.page, this.page.title)]), // Navigation popover
    vue.h(Popover_vue_vue_type_style_index_0_lang.s, {
      id: this.navPopoverId,
      contentClass: 'vc-nav-popover-container'
    }, {
      // Navigation pane
      default: function _default() {
        return vue.h(CalendarNav.s, {
          value: _this.page,
          validator: _this.canMove,
          onInput: function onInput($event) {
            return _this.move($event);
          }
        }, helpers.m({}, _this.$slots));
      }
    })])])]); // Weekdays

    var weekdays = vue.h(Grid_vue_vue_type_style_index_0_lang.s, {
      class: 'vc-weekdays',
      items: this.weekdayLabels,
      rows: 1,
      columns: 7,
      columnWidth: '1fr',
      disableFocus: true
    }, {
      cell: function cell(_ref) {
        var wl = _ref.item;
        return vue.h('div', {
          key: wl,
          class: 'vc-weekday'
        }, [wl]);
      }
    }); // Weeks

    var weeks = vue.h(Grid_vue_vue_type_style_index_0_lang.s, {
      class: 'vc-days',
      items: this.page.days,
      rows: this.page.weeks,
      columns: 7,
      columnWidth: '1fr',
      disableFocus: true
    }, {
      cell: function cell(_ref2) {
        var day = _ref2.item;
        return vue.h(CalendarDay_vue_vue_type_style_index_0_lang.a, helpers.m(helpers.m({}, _this.$attrs), {}, {
          day: day,
          slots: _this.$slots
        }));
      }
    });
    return vue.h('div', {
      class: 'vc-pane',
      ref: 'pane'
    }, [header, vue.h('div', {
      class: 'vc-weeks'
    }, [weekdays, weeks])]);
  },
  props: {
    page: Object,
    titlePosition: String,
    navVisibility: String,
    canMove: {
      type: Function,
      default: function _default() {
        return true;
      }
    }
  },
  data: function data() {
    return {
      navPopoverId: helpers.n()
    };
  },
  computed: {
    navVisibility_: function navVisibility_() {
      return this.propOrDefault('navVisibility', 'navVisibility');
    },
    navPlacement: function navPlacement() {
      switch (this.titlePosition) {
        case 'left':
          return 'bottom-start';

        case 'right':
          return 'bottom-end';

        default:
          return 'bottom';
      }
    },
    navPopoverEvents: function navPopoverEvents() {
      return CalendarDay_vue_vue_type_style_index_0_lang.g({
        id: this.navPopoverId,
        visibility: this.navVisibility_,
        placement: this.navPlacement,
        modifiers: [{
          name: 'flip',
          options: {
            fallbackPlacements: ['bottom']
          }
        }],
        isInteractive: true,
        isRenderFn: true
      });
    },
    weekdayLabels: function weekdayLabels() {
      var _this2 = this;

      return this.locale.getWeekdayDates().map(function (d) {
        return _this2.format(d, _this2.masks.weekdays);
      });
    }
  },
  methods: {
    move: function move(page) {
      this.$emit('update:page', page);
    }
  }
};var css_248z = ".vc-pane {\r\n  -webkit-flex-grow: 1;\r\n      -ms-flex-positive: 1;\r\n          flex-grow: 1;\r\n  -ms-flex-negative: 1;\r\n          -webkit-flex-shrink: 1;\r\n          flex-shrink: 1;\r\n  -webkit-align-self: flex-start;\r\n      -ms-flex-item-align: start;\r\n          align-self: flex-start;\r\n  display: -webkit-flex;\r\n  display: -ms-flexbox;\r\n  display: flex;\r\n  -webkit-flex-direction: column;\r\n      -ms-flex-direction: column;\r\n          flex-direction: column;\r\n  -webkit-justify-content: center;\r\n      -ms-flex-pack: center;\r\n          justify-content: center;\r\n  -webkit-align-items: stretch;\r\n      -ms-flex-align: stretch;\r\n          align-items: stretch;\n}\n.vc-horizontal-divider {\r\n  -webkit-align-self: center;\r\n      -ms-flex-item-align: center;\r\n          align-self: center;\n}\n.vc-header {\r\n  -ms-flex-negative: 0;\r\n          -webkit-flex-shrink: 0;\r\n          flex-shrink: 0;\r\n  display: -webkit-flex;\r\n  display: -ms-flexbox;\r\n  display: flex;\r\n  -webkit-align-items: stretch;\r\n      -ms-flex-align: stretch;\r\n          align-items: stretch;\r\n  color: var(--gray-900);\r\n  -webkit-user-select: none;\r\n      -ms-user-select: none;\r\n          user-select: none;\r\n  padding: 10px 10px 0 10px;\n}\n.vc-header.align-left {\r\n    -webkit-order: -1;\r\n        -ms-flex-order: -1;\r\n            order: -1;\r\n    -webkit-justify-content: flex-start;\r\n        -ms-flex-pack: start;\r\n            justify-content: flex-start;\n}\n.vc-header.align-right {\r\n    -webkit-order: 1;\r\n        -ms-flex-order: 1;\r\n            order: 1;\r\n    -webkit-justify-content: flex-end;\r\n        -ms-flex-pack: end;\r\n            justify-content: flex-end;\n}\n.vc-title-layout {\r\n  display: -webkit-flex;\r\n  display: -ms-flexbox;\r\n  display: flex;\r\n  -webkit-justify-content: center;\r\n      -ms-flex-pack: center;\r\n          justify-content: center;\r\n  -webkit-align-items: center;\r\n      -ms-flex-align: center;\r\n          align-items: center;\r\n  -webkit-flex-grow: 1;\r\n      -ms-flex-positive: 1;\r\n          flex-grow: 1;\n}\n.vc-title-layout.align-left {\r\n    -webkit-justify-content: flex-start;\r\n        -ms-flex-pack: start;\r\n            justify-content: flex-start;\n}\n.vc-title-layout.align-right {\r\n    -webkit-justify-content: flex-end;\r\n        -ms-flex-pack: end;\r\n            justify-content: flex-end;\n}\n.vc-title-wrapper {\r\n  position: relative;\n}\n.vc-title {\r\n  font-size: var(--text-lg);\r\n  color: var(--gray-800);\r\n  font-weight: var(--font-semibold);\r\n  cursor: pointer;\r\n  -webkit-user-select: none;\r\n      -ms-user-select: none;\r\n          user-select: none;\r\n  white-space: nowrap;\r\n  padding: 0 8px;\r\n  line-height: 27px;\n}\n.vc-title:hover {\r\n    opacity: 0.75;\n}\n.vc-weekday {\r\n  display: -webkit-flex;\r\n  display: -ms-flexbox;\r\n  display: flex;\r\n  -webkit-justify-content: center;\r\n      -ms-flex-pack: center;\r\n          justify-content: center;\r\n  -webkit-align-items: center;\r\n      -ms-flex-align: center;\r\n          align-items: center;\r\n  -webkit-flex: 1;\r\n      -ms-flex: 1;\r\n          flex: 1;\r\n  color: var(--gray-500);\r\n  font-size: var(--text-sm);\r\n  font-weight: var(--font-bold);\r\n  margin: 4px 0 6px 0;\r\n  cursor: default;\r\n  -webkit-user-select: none;\r\n      -ms-user-select: none;\r\n          user-select: none;\n}\n.vc-weeks {\r\n  -ms-flex-negative: 1;\r\n          -webkit-flex-shrink: 1;\r\n          flex-shrink: 1;\r\n  -webkit-flex-grow: 1;\r\n      -ms-flex-positive: 1;\r\n          flex-grow: 1;\r\n  padding: 5px 6px 7px 6px;\n}\n.vc-nav-popover-container {\r\n  color: var(--white);\r\n  font-size: var(--text-sm);\r\n  font-weight: var(--font-semibold);\r\n  background-color: var(--gray-800);\r\n  border: 1px solid;\r\n  border-color: var(--gray-700);\r\n  border-radius: var(--rounded-lg);\r\n  padding: 4px;\r\n  box-shadow: var(--shadow);\n}\n.vc-is-dark .vc-header {\r\n    color: var(--gray-200);\n}\n.vc-is-dark .vc-title {\r\n    color: var(--gray-100);\n}\n.vc-is-dark .vc-weekday {\r\n    color: var(--accent-200);\n}\n.vc-is-dark .vc-nav-popover-container {\r\n    color: var(--gray-800);\r\n    background-color: var(--white);\r\n    border-color: var(--gray-100);\n}\n";
styleInject_es.s(css_248z);exports.s=script;
